src/pydvl/valuation/methods/gt_shapley.py

```diff
@@ -12,325 +12,375 @@
+    n_samples = np.log(n_obs * (n_obs - 1) / delta)
+    n_samples /= 1 - q_tot
+    n_samples /= _h(epsilon / (2 * Z * np.sqrt(n_obs) * (1 - q_tot)))
 
+    return int(n_samples)
 
-def num_samples_eps_delta(
-    eps: float, delta: float, n: int, utility_range: float
-) -> int:
-    r"""Implements the formula in Theorem 3 of (Jia, R. et al., 2019)<sup><a href="#jia_efficient_2019">1</a></sup>
-    which gives a lower bound on the number of samples required to obtain an
-    (ε/√n,δ/(N(N-1))-approximation to all pair-wise differences of Shapley
-    values, wrt. $\ell_2$ norm.
 
-    Args:
-        eps: ε
-        delta: δ
-        n: Number of data points
-        utility_range: Range of the [Utility][pydvl.utils.utility.Utility] function
-    Returns:
-        Number of samples from $2^{[n]}$ guaranteeing ε/√n-correct Shapley
-            pair-wise differences of values with probability 1-δ/(N(N-1)).
+class GroupTestingProblem(NamedTuple):
+    """Solver agnostic representation of the group-testing problem."""
+
+    utility_differences: NDArray[np.float_]
+    total_utility: float
+    epsilon: float
 
-    !!! tip "New in version 0.4.0"
+
+class GTSampler(StochasticSamplerMixin, IndexSampler):
+    """Sampler for the group-testing algorithm.
+
+    Methods that are specific to semi-values like weight and make_strategy are not
+    implemented.
+
+    Args:
+        batch_size: The number of samples to draw in each batch.
+        seed: Seed for the random number generator.
 
     """
-    constants = _constants(n=n, epsilon=eps, delta=delta, utility_range=utility_range)
-    return int(constants.T)
 
+    def __init__(self, batch_size: int = 1, seed: Seed | None = None):
+        super().__init__(batch_size=batch_size, seed=seed)
 
-def _group_testing_shapley(
-    u: Utility,
-    n_samples: int,
-    progress: bool = False,
-    job_id: int = 1,
-    seed: Optional[Union[Seed, SeedSequence]] = None,
-):
-    """Helper function for
-    [group_testing_shapley()][pydvl.valuation.methods.gt_shapley.group_testing_shapley].
+    def _generate(self, indices: IndexSetT) -> SampleGenerator:
+        n_obs = len(indices)
+        sample_sizes = _create_sample_sizes(n_obs)
+        probabilities = _create_sampling_probabilities(sample_sizes)
 
-    Computes utilities of sets sampled using the strategy for estimating the
-    differences in Shapley values.
+        while True:
+            size = self._rng.choice(sample_sizes, p=probabilities)
+            subset = random_subset_of_size(indices, size=size, seed=self._rng)
+            yield Sample(idx=None, subset=subset)
+
+    @staticmethod
+    def weight(n: int, subset_len: int) -> float:
+        raise NotImplementedError("This is not a semi-value sampler.")
+
+    def make_strategy(
+        self,
+        utility: UtilityBase,
+        coefficient: Callable[[int, int], float] | None = None,
+    ) -> EvaluationStrategy:
+        raise NotImplementedError("This is not a semi-value sampler.")
```

Comment:
This will disappear in a later refactoring, correct?

Reply:
Yes, exactly.